b2/b2.py

```diff
@@ -1574,6 +1574,308 @@ def download_file_by_id_helper(
+        """
+        Only for local folders, returns the full path to the file.
+        """
+        raise NotImplementedError()
+
+
+class LocalFolder(Folder):
+    """
+    Folder interface to a directory on the local machine.
+    """
+
+    def __init__(self, root):
+        self.root = os.path.abspath(root)
+        self.relative_paths = self._get_all_relative_paths(self.root)
+
+    def folder_type(self):
+        return 'local'
+
+    def all_files(self):
+        for relative_path in self.relative_paths:
+            yield self._make_file(relative_path)
+
+    def make_full_path(self, file_name):
+        return os.path.join(self.root, file_name.replace('/', os.path.sep))
+
+    def _get_all_relative_paths(self, root_path):
+        """
+        Returns a sorted list of all of the files under the given root,
+        relative to that root
+        """
+        result = []
+        for dirpath, dirnames, filenames in os.walk(root_path):
+            for filename in filenames:
+                full_path = os.path.join(dirpath, filename)
+                relative_path = full_path[len(root_path) + 1:]
+                result.append(relative_path)
+        return sorted(result)
+
+    def _make_file(self, relative_path):
+        full_path = os.path.join(self.root, relative_path)
+        mod_time = os.path.getmtime(full_path)
+        slashes_path = '/'.join(relative_path.split(os.path.sep))
+        version = FileVersion(full_path, mod_time, "upload")
+        return File(slashes_path, [version])
+
+
+def next_or_none(iterator):
+    """
+    Returns the next item from the iterator, or None if there are no more.
+    """
+    try:
+        return six.advance_iterator(iterator)
+    except StopIteration:
+        return None
+
+
+def zip_folders(folder_a, folder_b):
+    """
+    An iterator over all of the files in the union of two folders,
+    matching file names.
+
+    Each item is a pair (file_a, file_b) with the corresponding file
+    in both folders.  Either file (but not both) will be None if the
+    file is in only one folder.
+    :param folder_a: A Folder object.
+    :param folder_b: A Folder object.
+    """
+    iter_a = folder_a.all_files()
+    iter_b = folder_b.all_files()
+    current_a = next_or_none(iter_a)
+    current_b = next_or_none(iter_b)
+    while current_a is not None or current_b is not None:
+        if current_a is None:
+            yield (None, current_b)
+            current_b = next_or_none(iter_b)
+        elif current_b is None:
+            yield (current_a, None)
+            current_a = next_or_none(iter_a)
+        elif current_a.name < current_b.name:
+            yield (current_a, None)
+            current_a = next_or_none(iter_a)
+        elif current_b.name < current_a.name:
+            yield (None, current_b)
+            current_b = next_or_none(iter_b)
+        else:
+            assert current_a.name == current_b.name
+            yield (current_a, current_b)
```

Comment:
I don't think so.  In the case where `current_a` has file `a.txt` and `current_b` has `b.txt`, it needs to yield `(current_a, None)`